src/Compilers/Core/Portable/Compilation.EmitStreamProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// The compilation object is an immutable representation of a single invocation of the
+    /// compiler. Although immutable, a compilation is also on-demand, and will realize and cache
+    /// data as necessary. A compilation can produce a new compilation from existing compilation
+    /// with the application of small deltas. In many cases, it is more efficient than creating a
+    /// new compilation from scratch, as the new compilation can reuse information from the old
+    /// compilation.
+    /// </summary>
+    public abstract partial class Compilation
+    {
+        /// <summary>
+        /// Abstraction that allows the caller to delay the creation of the <see cref="Stream"/> values 
+        /// until they are actually needed.
+        /// </summary>
+        internal abstract class EmitStreamProvider
```

Comment:
I am not sure if we should bundle different streams together. Probably each stream, including XML doc stream, should have it's own provider. If we want to decrease a number of parameters passed around, we could use a structure to combine them. If consumer gives a non-null StremProvider for a certain stream, than that means that the stream "exists", which eliminates the need for HasPdbStream property. 
Also, It would be good if the compiler could close a stream. For example, if we could close XML Doc stream as soon as we are done writing to it, we wouldn't have a problem with embedding it as a resource later. Perhaps StreamPrivider should have a special method for closing a stream, if client needs to do something special at this point.